theano/tensor/nnet/ctc.py

```diff
@@ -0,0 +1,148 @@
+        the minibatch.
+
+    Returns
+    -------
+    1-D tensor
+        Cost of each example in the minibatch. Tensor is of shape
+        (time index, minibatch index, probabilities).
+    """
+    __props__ = ('compute_grad',)
+
+    func_file = "./ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_cpu)"
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_cpu_no_grad)"
+
+        gof.COp.__init__(self, self.func_file, self.func_name)
+        gof.OpenMPOp.__init__(self)
+
+        self.compute_grad = compute_grad
+
+        if config.ctc.root == "":
+            raise ValueError('ctc.root variable is not set, please set it '
+                             'to the root directory of the CTC library in '
+                             'your system.')
+
+    def c_lib_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the compiled library (libwarpctc.so) is available
+            # at the build directory of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "build"))
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc"]
+
+    def c_header_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_compile_args(self):
+        return gof.OpenMPOp.c_compile_args(self)
+
+    def c_headers(self):
+        return ["ctc.h"] + gof.OpenMPOp.c_headers(self)
+
+    def make_node(self, activations, labels, input_lengths):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'ConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        t_activations = T.as_tensor_variable(activations)
+        # Ensure activations array is C-contiguous
+        t_activations = cpu_contiguous(t_activations)
+
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
+        if t_activations.type.dtype != 'float32':
+            raise TypeError('Activations must use the float32 type!')
+
+        if t_labels.type.dtype != 'int32':
+            raise TypeError('Labels must use the int32 type!')
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
+        costs = T.fvector(name="ctc_cost")
+
+        if self.compute_grad:
+            gradients = T.ftensor3(name="ctc_grad")
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
+        return gof.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                         outputs=[costs, gradients])
```

Comment:
Here also, build outputs list in same way as you do on GPU, ie:
```python
# pseudocode
outputs = [costs]
if compute_grad:
   outputs += [gradients]
```

(as gradients may not exist if compute_grad is false, It is important for Op consistency and reusability !).